stripe/src/test/java/com/stripe/android/IssuingCardPinServiceTest.java

```diff
@@ -0,0 +1,154 @@
+                mRequestExecutor,
+                false
+        );
+
+        service = new IssuingCardPinService(
+                ephemeralKeyProvider, apiHandler);
+    }
+
+    @Test
+    public void testRetrieval()
+            throws InvalidRequestException, APIConnectionException {
+
+        StripeResponse response =
+                new StripeResponse(200,
+                        "{\"card\":\"ic_abcdef\",\"pin\":\"1234\"}",
+                        null);
+
+        when(mRequestExecutor.execute(
+                eq(StripeApiHandler.RestMethod.GET),
+                eq("https://api.stripe.com/v1/issuing/cards/ic_abcdef/pin"),
+                anyMapOf(String.class, Object.class),
+                any(RequestOptions.class)
+        )).thenReturn(response);
+
+        IssuingCardPinService.IssuingCardPinRetrievalListener mockListener =
+                mock(IssuingCardPinService.IssuingCardPinRetrievalListener.class);
+
+        service.retrievePin(
+                "ic_abcdef",
+                "iv_abcd",
+                "123-456",
+                mockListener);
+
+        verify(mockListener).onIssuingCardPinRetrieved("1234");
+    }
+
+    @Test
+    public void testUpdate()
+            throws InvalidRequestException, APIConnectionException {
+
+        StripeResponse response =
+                new StripeResponse(200,
+                        "{\"card\":\"ic_abcdef\",\"pin\":\"\"}",
+                        null);
+
+        when(mRequestExecutor.execute(
+                eq(StripeApiHandler.RestMethod.POST),
+                eq("https://api.stripe.com/v1/issuing/cards/ic_abcdef/pin"),
+                anyMapOf(String.class, Object.class),
+                any(RequestOptions.class)
+        )).thenReturn(response);
+
+        IssuingCardPinService.IssuingCardPinUpdateListener mockListener =
+                mock(IssuingCardPinService.IssuingCardPinUpdateListener.class);
+
+        service.updatePin(
+                "ic_abcdef",
+                "1234",
+                "iv_abcd",
+                "123-456",
+                mockListener);
+
+        verify(mockListener).onIssuingCardPinUpdated();
+    }
+
+    @Test
+    public void testRetrievalFailsWithReason()
+            throws InvalidRequestException, APIConnectionException {
+
+        StripeResponse response =
+                new StripeResponse(400,
+                        "{\"error\": {\"code\": \"incorrect_code\", \"message\": " +
+                                "\"Verification failed\", \"type\": \"invalid_request_error\"}}",
+                        null);
+
+        when(mRequestExecutor.execute(
+                eq(StripeApiHandler.RestMethod.GET),
+                eq("https://api.stripe.com/v1/issuing/cards/ic_abcdef/pin"),
+                anyMapOf(String.class, Object.class),
+                any(RequestOptions.class)
+        )).thenReturn(response);
+
+        IssuingCardPinService.IssuingCardPinRetrievalListener mockListener =
```

Comment:
use `@Mock` annotation

Reply:
oh, ok 👍

Reply:
hmm, `Mock is not available for local variables`

Reply:
make it an instance variable

Reply:
👍